lib/Nodes/TableNode.php

```diff
@@ -4,177 +4,436 @@
     {
-        $line = utf8_decode($line);
+        if ($this->isCompiled === true) {
+            throw new LogicException('Cannot push data after TableNode is compiled');
+        }
 
-        if ($parts !== null) {
-            // New line in the tab
-            if ($parts[2] !== $this->parts[2]) {
-                return false;
-            }
+        $this->separatorLineConfigs[$this->currentLineNumber] = $separatorLineConfig;
+        $this->currentLineNumber++;
+    }
+
+    public function pushContentLine(string $line) : void
+    {
+        if ($this->isCompiled === true) {
+            throw new LogicException('Cannot push data after TableNode is compiled');
+        }
+
+        $this->rawDataLines[$this->currentLineNumber] = utf8_decode($line);
+        $this->currentLineNumber++;
+    }
+
+    public function finalize(Parser $parser) : void
+    {
+        if ($this->isCompiled === false) {
+            $this->compile();
+        }
 
-            if ($this->type === self::TYPE_PRETTY) {
-                if ($parts[0] === true) {
-                    $this->headers[count($this->data) - 1] = true;
+        $tableAsString = $this->getTableAsString();
+
+        if (count($this->errors) > 0) {
+            $parser->getEnvironment()
+                ->getErrorManager()
+                ->error(sprintf("%s\nin file %s\n\n%s", $this->errors[0], $parser->getFilename(), $tableAsString));
+
+            $this->data = [];
+
+            return;
+        }
+
+        foreach ($this->data as $i => $row) {
+            foreach ($row->getColumns() as $col) {
+                // TODO - not sure if this ever did anything
+                // can't see how it would have been hit
+                /*
+                if ($rendered instanceof Node) {
+                    $rendered = $rendered->render();
+                }
+                */
+
+                $lines = explode("\n", $col->getContent());
+
+                if ($this->lineChecker->isListLine($lines[0], false)) {
+                    $node = $parser->parseFragment($col->getContent())->getNodes()[0];
+                } else {
+                    $node = $parser->createSpanNode($col->getContent());
                 }
-                $this->data[] = [];
-            } elseif (count($this->headers) === 0) {
-                $this->headers = array_fill_keys(array_keys($this->data), true);
+
+                $col->setNode($node);
             }
+        }
+    }
+
+    /**
+     * Looks at all the raw data and finally populates the data
+     * and headers.
+     */
+    private function compile() : void
+    {
+        $this->isCompiled = true;
+
+        if ($this->type === self::TYPE_SIMPLE) {
+            $this->compileSimpleTable();
         } else {
-            if ($this->type === self::TYPE_SIMPLE) {
-                $this->data[] = [];
+            $this->compilePrettyTable();
+        }
+    }
+
+    private function compileSimpleTable() : void
```

Comment:
The old logic for turning the raw config/lines into a meaningful `$this->data` property were completely removed. I couldn't make any sense of them. I've replaced them with 2 entirely separate functions for the "simple table" and the "pretty/grid table". Each of these is super unfancy, because I want them to be readable and because algorithms are hard for me :).